static_frame/core/index_hierarchy.py

```diff
@@ -1794,70 +2364,173 @@ class IndexHierarchyGO(IndexHierarchy):
     STATIC = False
 
     _IMMUTABLE_CONSTRUCTOR = IndexHierarchy
-    _LEVEL_CONSTRUCTOR = IndexLevelGO
     _INDEX_CONSTRUCTOR = IndexGO
 
     __slots__ = (
-            '_levels', # IndexLevel
+            '_indices',
+            '_indexers',
+            '_name',
             '_blocks',
+            '_values',
             '_recache',
-            '_name'
+            '_index_type',
             )
 
-    _levels: IndexLevelGO
+    _indices: tp.List[IndexGO] # type: ignore
 
-    def append(self, value: tp.Sequence[tp.Hashable]) -> None:
+    def append(self: IHGO,
+            value: tp.Sequence[tp.Hashable],
+            ) -> None:
         '''
         Append a single label to this index.
         '''
-        self._levels.append(value)
-        self._recache = True
+        # TODO: How to handle corrupted instance state if an exception is raised?
+        if value in self:
+            raise ErrorInitIndexNonUnique(
+                f"The label '{value}' is already in the index."
+            )
+
+        label_indexers = []
+
+        for depth, label_at_depth in enumerate(value):
+            if label_at_depth in self._indices[depth]:
+                label_index = self._indices[depth]._loc_to_iloc(label_at_depth)
+            else:
+                label_index = len(self._indices[depth])
+                self._indices[depth].append(label_at_depth)
 
-    def extend(self, other: IndexHierarchy) -> None:
+            label_indexers.append(label_index)
+
+        for i, label_index in enumerate(label_indexers):
+            self._indexers[i] = np.append(self._indexers[i], label_index)
```

Comment:
Up until now, `np.append` has never been used in StaticFrame and is an anti-pattern in some regards. This is equivalent to an np.concatenate, and forces a fully copy of the underlying arrays for each addition. Please evaluate the performance of adding multiple single columns in a `FrameGO` with `IndexHierarchyGO` as columns with this implementation. I wonder if a cached approach, using a mutable list that defers creating an array as we do in `IndexGO`, might be better.

Reply:
I was actually talking with Brandt this morning and learned for the first time what the purpose of `_recache`` and `_update_array_cache` was used for!

I will spend time thinking about how to implement this here :+1: